Update daemon JVM documentation (#29380)
gradle/gradle: platforms/documentation/docs/src/docs/userguide/optimizing-performance/gradle_daemon.adoc

To find the PIDs for all Daemons regardless of Gradle version, see <<gradle_daemon.adoc#find_all_daemons,Find Daemons>>.

[[sec:daemon_jvm_criteria]]
== Configuring the JVM to be used

NOTE: Daemon JVM discovery and criteria are <<feature_lifecycle.adoc#sec:incubating_state,incubating>> features and are subject to change in a future release.

By default, the Gradle daemon runs on the same JVM installation that started the build.
Gradle defaults to the current shell path and `JAVA_HOME` environment variable to locate a usable JVM.

Alternatively, you could specify a different JVM installation for the build using the <<build_environment.adoc#sec:gradle_configuration_properties,`org.gradle.java.home` Gradle property>> or programmatically through the Tooling API.

Building on the <<toolchains.adoc#toolchains,toolchain feature>>, you can now use declarative criteria to specify the JVM requirements for the build.

=== Daemon JVM criteria

> The _daemon JVM criteria_ feature is built based on a task, similar to the <<gradle_wrapper.adoc#customizing_wrapper,`wrapper` one>>.
> The task uses its configuration in the build script or command-line arguments to capture a JVM definition in the `gradle/daemon-jvm.properties` file.

Reviewer: ```suggestion
The _daemon JVM criteria_ is controlled by a task, similarly to how <<gradle_wrapper.adoc#customizing_wrapper,`wrapper` task>> updates the wrapper properties.
When the task runs, it creates or updates the criteria in the `gradle/daemon-jvm.properties` file.
For more control, the task can be further configured in the build script or via command-line arguments.
```

> The task, when run, will use its configuration, found in build script or as command line arguments, and will capture a JVM definition in the `gradle/daemon-jvm.properties` file.

Reviewer 2: ```suggestion
The task uses its configuration in the build script or command-line arguments to capture a JVM definition in the `gradle/daemon-jvm.properties` file.
```

As with the wrapper, the generated file should be checked into version control.
This will ensure any developer or CI server that runs the build will use the same JVM version.

With the following configuration, running `./gradlew updateDaemonJvm` will generate the file shown below:
====
include::sample[dir="snippets/daemon-jvm/customized-task/kotlin",files="build.gradle.kts[tags=customized-daemon-jvm-task]"]
include::sample[dir="snippets/daemon-jvm/customized-task/groovy",files="build.gradle[tags=customized-daemon-jvm-task]"]
====

====
.gradle/daemon-jvm.properties
[source,properties]
----

> #This file is generated by updateDaemonJvm

Reviewer: ```suggestion
# This file is generated by updateDaemonJvm
```

Author: Yes, we should fix that, but that's a code change. I copied it from a real generated task

toolchainVersion=17
----
====

> Without editing the build script, you could produce the same output using the following command:

Reviewer: ```suggestion
The same properties file can be produced without configuring the task in the build script, and using a command-line argument instead:
```

----
$ ./gradlew updateDaemonJvm --jvm-version=17
----

> If you run the task without any argument, then the JVM used by the daemon will provide the version value.

Reviewer: ```suggestion
If you run the task without any arguments and the properties file does not exist, then the JVM used by the daemon will provide the version value.
```
